### ISSUE TITLE
mlx_whisper memory usage keeps growing for every new audio chunk processed with word_timestamps on.

### ISSUE THREAD

**Issue by gencerege** [ml-explore/mlx-examples#1254]
I’m integrating mlx_whisper/whisper_streaming with a custom VAD for local usage. My goal is to run live transcription for an arbitrarily long duration, and whisper_streaming requires word_timestamps. However, I’ve noticed that for each new chunk of audio processed from the microphone input, the application’s memory usage grows continuously.

This issue seems to be directly related to the mlx_whisper.transcribe() function when word_timestamps is enabled. It does not occur if:

    - You process the same audio chunk repeatedly.
    - You run the code with word_timestamps = False.

Below is a minimal code snippet to reproduce the issue. If you use an audio file longer than two minutes with word_timestamps=True, memory usage will increase with each iteration (approximately 10MB, updates on the activity monitor can vary but its a lot even if it is not exactly 10MB):
```
import mlx_whisper

full_audio = load_audio('your_audio.wav')
for l in range(20): # this is to make sure every iteration a new audio chunk is processed by introducing an offset.
    for i in range(0, 5):
        # MLX Only
        text = mlx_whisper.transcribe(full_audio[i*40*8000 + l*1000:(i+1)*40*8000 + l*1000], path_or_hf_repo="whisper-v3-turbo", word_timestamps = True)
```

I got down to the find_alignment function and the line logits, cross_qk = model.forward_with_cross_qk(mel[None, :], tokens[None, :]), which seemed like the only possibility that could be adding large overhead to the model object in a way that increases RAM constantly. Could not delve any more as sadly I do not know any c or cpp (although I have searched for the function quite a bit on the repo, wasnt able to find it). Also, I am a beginner in python so I might be completely wrong here. 
Thank you!

Python version: 3.12.8 on Macbook Pro M2 Max

**Comment by awni** [comment]
I think this is expected cache growth. The word timestamps exacerbates it because the variable shapes of the qk scores.

You can control the memory growth using the [metal memory API](https://ml-explore.github.io/mlx/build/html/python/metal.html). The simplest option is to set a cache limit using `mx.metal.set_cache_limit`. Maybe try 100MB or so. The tradeoff is if you set it too low then the program could slow down, so give it as much room as you are comfortable with.

Another option is to manually clear the cache every now and then `mx.metal.clear_cache`. If you do it every few chunks of audio then it should limit cache growth with minimal performance implications.

**Comment by awni** [comment]
(let me know if that doesn't solve your issue and I will reopen this).

**Comment by gencerege** [comment]
Thank you very much, setting the cache limit solved my issue perfectly with no observable performance loss.

**Comment by AngeloSavich** [comment]
Hi @awni — following up on the `word_timestamps=True` memory growth (#1254, likely same root cause as #1366).

**Root cause**: `find_alignment()` produces variable-shape cross-attention tensors each chunk. These create uniquely-sized Metal allocations that can't be reused from the buffer pool, so the pool grows unbounded (~50-100MB per 30s chunk).

**Fix**: `mx.clear_cache()` after each chunk's word-timestamp alignment. This flushes the Metal buffer pool of those one-off allocations. It's the same pattern already used in `mlx-lm/generate.py` (clear_cache every 256 tokens) and consistent with your performance guide ([gist](https://gist.github.com/awni/4beb1f7dfefc6f9426f3a7deee74af50)).

**Test results** (49 files / 41.5 hours, Rode WirelessPRO, M4 Mac):

| Variant | Metal cache | Wall time/file |
|---------|------------|----------------|
| Baseline (no fix) | 9,092 MB | ~32s |
| `clear_cache()` per chunk | 61 MB | ~32s |
| `clear_cache()` per file | 61 MB | ~29s (fastest — eliminates pool search overhead) |

99.3% cache reduction with no performance penalty. Per-file clearing is actually slightly faster because the allocator doesn't have to search a bloated pool.

The minimal fix is a one-line `mx.clear_cache()` call after `add_word_timestamps()` returns in the transcription loop.

I have a fork with the fix at `AngeloSavich/mlx-examples` branch `fix/whisper-word-timestamps-memory` — happy to open a PR if you're interested.
